src/main/kotlin/com/ekino/oss/recaptcha/config/ReCaptchaConfiguration.kt

```diff
@@ -0,0 +1,82 @@
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.databind.SerializationFeature
+import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import mu.KotlinLogging
+import okhttp3.OkHttpClient
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication.Type.SERVLET
+import org.springframework.boot.context.properties.EnableConfigurationProperties
+import org.springframework.context.annotation.Bean
+import org.springframework.context.annotation.Configuration
+import org.springframework.context.annotation.Import
+import org.springframework.http.HttpMethod
+import retrofit2.Retrofit
+import retrofit2.converter.jackson.JacksonConverterFactory
+
+private val logger = KotlinLogging.logger {}
+
+/**
+ * Configuration for reCaptcha filter.
+ * Available only for a servlet-based application.
+ * Disabled if {@code security.recaptcha.enabled} is false.
+ */
+@ConditionalOnWebApplication(type = SERVLET)
+@Configuration(proxyBeanMethods = false)
+@EnableConfigurationProperties(ReCaptchaProperties::class)
+@ConditionalOnProperty(value = ["security.recaptcha.enabled"], havingValue = "true", matchIfMissing = true)
+@Import(ReCaptchaValidationService::class)
+class ReCaptchaConfiguration(private val reCaptchaProperties: ReCaptchaProperties) {
+
+  private val objectMapper: ObjectMapper by lazy {
+    jacksonObjectMapper().apply {
+      setDefaultPropertyInclusion(JsonInclude.Include.NON_ABSENT)
+      setSerializationInclusion(JsonInclude.Include.NON_ABSENT)
+      configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
+      configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false)
+      findAndRegisterModules()
+    }
+  }
+
+  /**
+   * Filter registration, protecting only urls defined in properties.
+   */
+  @Bean
+  fun reCaptchaFilter(
+    reCaptchaValidationService: ReCaptchaValidationService,
+    reCaptchaFailureService: ReCaptchaFailureService
+  ) = ReCaptchaFilter(
+    reCaptchaValidationService,
+    reCaptchaProperties.responseName,
+    reCaptchaProperties.byPassKey,
+    reCaptchaProperties.urlPatterns,
+    reCaptchaProperties.filteredMethods.map(HttpMethod::name).toSet(),
+    reCaptchaFailureService)
+
+  @Bean
+  fun reCaptchaClient(): ReCaptchaClient =
+    with(reCaptchaProperties.client) {
+      Retrofit.Builder()
+        .baseUrl(url)
+        .addConverterFactory(JacksonConverterFactory.create(objectMapper))
+        .client(
+          OkHttpClient().newBuilder()
+            .connectTimeout(connectTimeout)
+            .readTimeout(readTimeout)
+            .writeTimeout(writeTimeout)
+            .build()
+        )
+        .build()
+        .create(ReCaptchaClient::class.java)
+    }
+
+  @Bean
+  fun reCaptchaFailureService() = ReCaptchaFailureService(objectMapper)
```

Comment:
Might there be a `@ConditionalOnMissingBean` missing ?

Reply:
More globally it might be a good idea to add tests akin to what is described in https://spring.io/blog/2018/03/07/testing-auto-configurations-with-spring-boot-2-0, for instance on overriding this bean and `ConditionalOnProperty`. This could be done here or as a technical issue

Reply:
Indeed, I forgot the `ConditionalOnMissingBean`, corrected.
For the tests, sure it will be better with thoses tests, I will add them in another commit :+1: